File: 2021/3.1.php
<?php


require "Helpers.php";
class DayThreePartOne
{
    function solve($filename)
    {
        $inputs = DayThreePartOne::prepareInputs($filename);
        $rates = $this->calculateRates($inputs);
        echo abs(bindec($rates["gamma"]) * bindec($rates["epsilon"])) . "\n";
    }

    static function prepareInputs($filename)
    {
        return array_map(function($line) {
            return array_map("intval", str_split($line));
        }, Helpers::splitInputs($filename));
    }

    function calculateRates($inputs)
    {
        $bits = [];
        foreach (range(0, count($inputs[0]) -1) as $index) {
            $bits[] = array_reduce($inputs, function($acc, $input) use(&$index) {
                $acc[] = $input[$index];
                return $acc;
            }, []);
        }

        $gamma = array_reduce($bits, function($acc, $input) {
            $max = array_count_values($input);
            return $acc.(string)array_keys($max, max($max))[0];
        }, "");
        $epsilon = array_reduce($bits, function($acc, $input) {
            $min = array_count_values($input);
            return $acc.(string)array_keys($min, min($min))[0];
        }, "");

        return [
            "gamma" => $gamma,
            "epsilon" => $epsilon
        ];
    }
}

$class = new DayThreePartOne();
// $class->solve("/2021/3test.txt");
$class->solve("/2021/3.txt");

File: 2020/03/3.1.rb
require_relative 'helpers.rb'

module DayThreePartOne
  module_function

  def solve(file)
    inputs = prepare_inputs(file)
    find_total_trees(inputs, 0, 0, 0)
  end

  def find_total_trees(inputs, total, x, y)
    row_length = inputs[0].length
    column_length = inputs.length
    x += 3
    x -= row_length if x >= row_length
    y += 1
    # y -= column_length if y >= column_length
    if y < column_length
      total += 1 if inputs[y][x] == '#'
      total = find_total_trees(inputs, total, x, y)
    end
    total
  end

  def prepare_inputs(file)
    Helpers
      .split_inputs_by_line(Helpers.read_file(file))
  end
end

# puts DayThreePartOne.solve(__dir__ + '/3test.txt')
puts DayThreePartOne.solve(__dir__ + '/3.txt')
